language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir

```diff
@@ -0,0 +1,100 @@
+// This file contains tests checking that type arguments used in phantom position
+// (phantom arguments) are not considered when deriving the abilities for a struct.
+
+module M1 {
+    struct NoAbilities { a: bool }
+}
+
+//! new-transaction
+
+module M2 {
+    import {{default}}.M1;
+
+    struct HasDrop<phantom T1, T2> has drop { a: bool }
+
+    // `WriteRef` requires drop
+    f(ref: &mut Self.HasDrop<M1.NoAbilities, M1.NoAbilities>) {
+        *move(ref) = HasDrop<M1.NoAbilities, M1.NoAbilities> { a : true };
+        return;
+    }
+}
+
+//! new-transaction
+
+module M3 {
+    import {{default}}.M1;
+
+    struct HasDrop<phantom T1, T2> has drop { a: bool }
+
+    // `Pop` requires drop
+    f() {
+        _ = HasDrop<M1.NoAbilities, M1.NoAbilities> { a: true };
+        return;
+    }
+ }
+
+ //! new-transaction
+
+ module M4 {
+    import {{default}}.M1;
+
+    struct HasDrop<phantom T1, T2> has drop { a: bool }
+
+    // Leaving value in local requires drop
+    f(x: Self.HasDrop<M1.NoAbilities, M1.NoAbilities>) {
+        return;
+    }
+ }
+
+ //! new-transaction
+
+ module M5 {
+    import {{default}}.M1;
+    struct HasCopy<phantom T1, T2> has copy { a : bool }
+
+    // `CopyLoc` requires copy
+    f(x: Self.HasCopy<M1.NoAbilities, M1.NoAbilities>): Self.HasCopy<M1.NoAbilities, M1.NoAbilities> * Self.HasCopy<M1.NoAbilities, M1.NoAbilities> {
+        return (copy(x), move(x));
+    }
+ }
+
+ //! new-transaction
+
+ module M6 {
```

Comment:
I don't see any with store. You should do something like `struct Cup<T: store> has key { value: T }` and `Cup<HasStore<NoAbilities>>` or even just `K has key { value: HasStore<NoAbilities>> }`

The generic one might be good since it will be checked at the operation, instead of the declaration

Reply:
Both tests are already considered. The first one in `abilities_constraints_[ok/bad].mvir` and the second in `fields_abilities_[ok/bad].mvir`.

Reply:
That is not quite the same test though, as it is not testing the interaction at the bytecode operation itself

Reply:
Ok, I added the test.